File: workers/python/kitchen_sink.py
from __future__ import annotations

import asyncio
from datetime import timedelta
from typing import Any, Awaitable, Callable, Coroutine, Optional, TypeVar, Union

import temporalio.workflow
from temporalio import exceptions, workflow
from temporalio.api.common.v1 import Payload
from temporalio.common import (
    Priority,
    RawValue,
    RetryPolicy,
    SearchAttributeKey,
    SearchAttributeUpdate,
)
from temporalio.converter import DefaultPayloadConverter
from temporalio.workflow import ActivityHandle, ChildWorkflowHandle

from protos.kitchen_sink_pb2 import (
    Action,
    ActionSet,
    ActivityCancellationType,
    AwaitableChoice,
    DoActionsUpdate,
    DoSignal,
    ExecuteActivityAction,
    WorkflowInput,
    WorkflowState,
)


@workflow.defn(name="kitchenSink")
class KitchenSinkWorkflow:
    action_set_queue: asyncio.Queue[ActionSet] = asyncio.Queue()
    workflow_state = WorkflowState()

    @workflow.signal
    async def do_actions_signal(self, signal_actions: DoSignal.DoSignalActions) -> None:
        if signal_actions.HasField("do_actions_in_main"):
            self.action_set_queue.put_nowait(signal_actions.do_actions_in_main)
        else:
            await self.handle_action_set(signal_actions.do_actions)

    @workflow.update
    async def do_actions_update(self, actions_update: DoActionsUpdate) -> Any:
        # IF variant was rejected we wouldn't even be in here, so access action set directly
        retval = await self.handle_action_set(actions_update.do_actions)
        if retval is not None:
            return retval
        return self.workflow_state

    @do_actions_update.validator
    def do_actions_update_val(self, actions_update: DoActionsUpdate):
        if actions_update.HasField("reject_me"):
            raise exceptions.ApplicationError("Rejected")

    @workflow.query
    def report_state(self, _: Any) -> WorkflowState:
        return self.workflow_state

    @workflow.run
    async def run(self, input: Optional[WorkflowInput] = None) -> Payload:
        workflow.logger.debug("Started kitchen sink workflow")

        if input and input.expected_signal_count > 0:
            raise exceptions.ApplicationError(
                "signal deduplication not implemented", non_retryable=True
            )

        # Run all initial input actions
        if input and input.initial_actions:
            for action_set in input.initial_actions:
                return_value = await self.handle_action_set(action_set)
                if return_value is not None:
                    return return_value

        # Run all actions from signals
        while True:
            action_set = await self.action_set_queue.get()
            return_value = await self.handle_action_set(action_set)
            if return_value is not None:
                return return_value

    async def handle_action_set(self, action_set: ActionSet) -> Optional[Payload]:
        return_value = None
        # If these are non-concurrent, just execute and return if requested
        if not action_set.concurrent:
            for action in action_set.actions:
                return_value = await self.handle_action(action)
                if return_value is not None:
                    return return_value
            return return_value

        # With a concurrent set, we'll create a task for each, only updating
        # return values if we should return, then awaiting on that or completion
        async def run_action(action: Action) -> None:
            maybe_return_value = await self.handle_action(action)

            if maybe_return_value is not None:
                nonlocal return_value
                return_value = maybe_return_value

        gather_fut = asyncio.gather(*[run_action(a) for a in action_set.actions])
        should_return_task = asyncio.create_task(
            workflow.wait_condition(lambda: return_value is not None)
        )
        # mypy cannot handle the heterogeneous arguments to `wait`
        done, _ = await workflow.wait(
            [gather_fut, should_return_task], return_when=asyncio.FIRST_COMPLETED
        )  # type: ignore
        for fut in done:
            await fut  # type: ignore
        return return_value

    async def handle_action(self, action: Action) -> Optional[Payload]:
        if action.HasField("return_result"):
            return action.return_result.return_this
        elif action.HasField("return_error"):
            raise exceptions.ApplicationError(action.return_error.failure.message)
        elif action.HasField("continue_as_new"):
            args = [RawValue(i) for i in action.continue_as_new.arguments]
            workflow.continue_as_new(args=args)
        elif action.HasField("timer"):
            await handle_awaitable_choice(
                asyncio.sleep(action.timer.milliseconds / 1000),
                action.timer.awaitable_choice,
            )
        elif action.HasField("exec_activity"):
            await handle_awaitable_choice(
                launch_activity(action.exec_activity),
                action.exec_activity.awaitable_choice,
            )
        elif action.HasField("exec_child_workflow"):
            child_action = action.exec_child_workflow
            child = child_action.workflow_type or "kitchenSink"
            args = [RawValue(i) for i in child_action.input]

            await handle_awaitable_choice(
                workflow.start_child_workflow(
                    child,
                    id=child_action.workflow_id,
                    args=args,
                    search_attributes=decode_search_attrs(
                        child_action.search_attributes, DefaultPayloadConverter()
                    ),
                ),
                child_action.awaitable_choice,
                after_started_fn=wait_task_complete,
                after_completed_fn=wait_child_wf_complete,
            )
        elif action.HasField("set_patch_marker"):
            if action.set_patch_marker.deprecated:
                workflow.deprecate_patch(action.set_patch_marker.patch_id)
                was_patched = True
            else:
                was_patched = workflow.patched(action.set_patch_marker.patch_id)

            if was_patched:
                return await self.handle_action(action.set_patch_marker.inner_action)
        elif action.HasField("set_workflow_state"):
            self.workflow_state = action.set_workflow_state
        elif action.HasField("await_workflow_state"):
            await workflow.wait_condition(
                lambda: self.workflow_state.kvs.get(action.await_workflow_state.key)
                == action.await_workflow_state.value
            )
        elif action.HasField("upsert_memo"):
            pass  # Python doesn't have memo upserting
        elif action.HasField("upsert_search_attributes"):
            updates: list[SearchAttributeUpdate[Any]] = []
            # TODO: Use RawValue after https://github.com/temporalio/sdk-python/issues/438
            #  and avoid checking key by name
            for k, v in action.upsert_search_attributes.search_attributes.items():
                if "Keyword" in k:
                    updates.append(
                        SearchAttributeKey.for_keyword(k).value_set(str(v.data[0]))
                    )
                else:
                    updates.append(SearchAttributeKey.for_int(k).value_set(v.data[0]))
            workflow.upsert_search_attributes(updates)
        elif action.HasField("nested_action_set"):
            return await self.handle_action_set(action.nested_action_set)
        elif action.HasField("nexus_operation"):
            raise exceptions.ApplicationError("ExecuteNexusOperation is not supported")
        else:
            raise exceptions.ApplicationError("unrecognized action: " + str(action))

        return None


def launch_activity(execute_activity: ExecuteActivityAction) -> ActivityHandle:
    act_type = "noop"
    args: list[Any] = []

    if execute_activity.HasField("delay"):
        act_type = "delay"
        args.append(execute_activity.delay)
    elif execute_activity.HasField("payload"):
        act_type = "payload"
        input_data = bytes(
            i % 256 for i in range(execute_activity.payload.bytes_to_receive)
        )
        args.append(input_data)
        args.append(execute_activity.payload.bytes_to_return)
    elif execute_activity.HasField("client"):
        act_type = "client"
        args.append(execute_activity.client)

    if execute_activity.HasField("is_local"):
        activity_task = workflow.start_local_activity(
            activity=act_type,
            args=args,
            schedule_to_close_timeout=timeout_or_none(
                execute_activity, "schedule_to_close_timeout"
            ),
            start_to_close_timeout=timeout_or_none(
                execute_activity, "start_to_close_timeout"
            ),
            schedule_to_start_timeout=timeout_or_none(
                execute_activity, "schedule_to_start_timeout"
            ),
            retry_policy=RetryPolicy.from_proto(execute_activity.retry_policy)
            if execute_activity.HasField("retry_policy")
            else None,
            # TODO: cancel type can be in local
        )
    else:
        if execute_activity.HasField("priority"):
            raise NotImplementedError("priority is not supported yet")

        activity_task = workflow.start_activity(
            activity=act_type,
            args=args,
            task_queue=execute_activity.task_queue,
            schedule_to_close_timeout=timeout_or_none(
                execute_activity, "schedule_to_close_timeout"
            ),
            start_to_close_timeout=timeout_or_none(
                execute_activity, "start_to_close_timeout"
            ),
            schedule_to_start_timeout=timeout_or_none(
                execute_activity, "schedule_to_start_timeout"
            ),
            heartbeat_timeout=timeout_or_none(execute_activity, "heartbeat_timeout"),
            retry_policy=RetryPolicy.from_proto(execute_activity.retry_policy)
            if execute_activity.HasField("retry_policy")
            else None,
            cancellation_type=convert_act_cancel_type(
                execute_activity.remote.cancellation_type
            ),
        )

    return activity_task


async def brief_wait(_: asyncio.Task):
    await asyncio.sleep(0.001)


async def wait_task_complete(task: asyncio.Task):
    await task


async def wait_child_wf_complete(task: asyncio.Task[ChildWorkflowHandle]):
    res = await task
    await res


async def handle_awaitable_choice(
    awaitable: Union[Coroutine, asyncio.Task],
    choice: AwaitableChoice,
    after_started_fn: Callable[[asyncio.Task], Awaitable] = brief_wait,
    after_completed_fn: Callable[[asyncio.Task], Awaitable] = wait_task_complete,
):
    if isinstance(awaitable, asyncio.Task):
        task = awaitable
    else:
        # Place the awaitable into a task so we can cancel it easily
        task = asyncio.create_task(awaitable)

    did_cancel = False
    try:
        if choice.HasField("abandon"):
            # Do nothing
            return
        elif choice.HasField("cancel_before_started"):
            task.cancel()
            did_cancel = True
            await task
        elif choice.HasField("cancel_after_started"):
            await after_started_fn(task)
            task.cancel()
            did_cancel = True
            await task
        elif choice.HasField("cancel_after_completed"):
            await after_completed_fn(task)
            task.cancel()
            did_cancel = True
        else:
            await task
    except asyncio.CancelledError:
        if not did_cancel:
            raise


# Various proto conversions below ==============================================


def timeout_or_none(
    activity_action: ExecuteActivityAction, timeout_field: str
) -> Optional[timedelta]:
    if activity_action.HasField(timeout_field):
        return timedelta(
            seconds=getattr(activity_action, timeout_field).seconds,
            microseconds=getattr(activity_action, timeout_field).nanos / 1000,
        )
    return None


def convert_act_cancel_type(
    ctype: ActivityCancellationType,
) -> temporalio.workflow.ActivityCancellationType:
    if ctype == ActivityCancellationType.TRY_CANCEL:
        return temporalio.workflow.ActivityCancellationType.TRY_CANCEL
    elif ctype == ActivityCancellationType.WAIT_CANCELLATION_COMPLETED:
        return temporalio.workflow.ActivityCancellationType.WAIT_CANCELLATION_COMPLETED
    elif ctype == ActivityCancellationType.ABANDON:
        return temporalio.workflow.ActivityCancellationType.ABANDON
    else:
        raise NotImplementedError("Unknown cancellation type " + str(ctype))


def decode_search_attrs(msg_map, converter):
    return {
        k: v if isinstance(v := converter.from_payload(p), list) else [v]
        for k, p in msg_map.items()
    }

File: workers/ruby/kitchen_sink.rb
require 'temporalio/workflow'
require 'temporalio/workflow/definition'
require_relative 'protos/kitchen_sink/kitchen_sink_pb'

KS = Temporal::Omes::KitchenSink

class KitchenSinkWorkflow < Temporalio::Workflow::Definition
  workflow_name 'kitchenSink'
  workflow_arg_hint KS::WorkflowInput
  workflow_result_hint Temporalio::Api::Common::V1::Payload

  def initialize
    super
    @action_set_queue = []
    @workflow_state = KS::WorkflowState.new
  end

  def execute(input = nil)
    Temporalio::Workflow.logger.debug('Started kitchen sink workflow')

    initial_return_value = nil
    if input&.initial_actions&.any?
      input.initial_actions.each do |action_set|
        return_value = handle_action_set(action_set)
        unless return_value.nil?
          initial_return_value = return_value
          break
        end
      end
    end

    if input&.expected_signal_count&.positive?
      raise Temporalio::Error::ApplicationError.new(
        'signal deduplication not implemented',
        non_retryable: true
      )
    end

    return initial_return_value unless initial_return_value.nil?

    loop do
      Temporalio::Workflow.wait_condition { !@action_set_queue.empty? }
      action_set = @action_set_queue.shift
      return_value = handle_action_set(action_set)
      return return_value unless return_value.nil?
    end
  end

  workflow_signal arg_hints: [KS::DoSignal::DoSignalActions]
  def do_actions_signal(signal_actions)
    if signal_actions.variant == :do_actions_in_main
      @action_set_queue << signal_actions.do_actions_in_main
    else
      handle_action_set(signal_actions.do_actions)
    end
  end

  workflow_update arg_hints: [KS::DoActionsUpdate]
  def do_actions_update(actions_update)
    retval = handle_action_set(actions_update.do_actions)
    return retval unless retval.nil?

    @workflow_state
  end

  workflow_update_validator :do_actions_update
  def do_actions_update_validator(actions_update)
    if actions_update.respond_to?(:reject_me) && actions_update.reject_me
      raise Temporalio::Error::ApplicationError, 'Rejected'
    end
  end

  workflow_query result_hint: KS::WorkflowState
  def report_state(_input = nil)
    @workflow_state
  end

  private

  def handle_action_set(action_set)
    return nil if action_set.nil?

    if action_set.concurrent
      return_value = nil
      futures = action_set.actions.map do |action|
        Temporalio::Workflow.async do
          result = handle_action(action)
          return_value = result unless result.nil?
        end
      end
      Temporalio::Workflow.wait_condition { futures.all?(&:done?) || !return_value.nil? }
      return_value
    else
      action_set.actions.each do |action|
        return_value = handle_action(action)
        return return_value unless return_value.nil?
      end
      nil
    end
  end

  def handle_action(action)
    case action.variant
    when :return_result
      return action.return_result.return_this
    when :return_error
      raise Temporalio::Error::ApplicationError, action.return_error.failure.message
    when :continue_as_new
      args = action.continue_as_new.arguments.map { |a| Temporalio::Converters::RawValue.new(a) }
      Temporalio::Workflow.continue_as_new(*args)
    when :timer
      handle_awaitable_choice(
        -> { Temporalio::Workflow.sleep(action.timer.milliseconds / 1000.0) },
        action.timer.awaitable_choice
      )
    when :exec_activity
      handle_awaitable_choice(
        -> { launch_activity(action.exec_activity) },
        action.exec_activity.awaitable_choice
      )
    when :exec_child_workflow
      child_action = action.exec_child_workflow
      child_type = child_action.workflow_type.empty? ? 'kitchenSink' : child_action.workflow_type
      args = child_action.input.map { |a| Temporalio::Converters::RawValue.new(a) }

      proto_sa = Temporalio::Api::Common::V1::SearchAttributes.new(
        indexed_fields: child_action.search_attributes.to_h
      )

      handle_awaitable_choice(
        lambda {
          Temporalio::Workflow.execute_child_workflow(
            child_type,
            *args,
            id: child_action.workflow_id,
            task_queue: child_action.task_queue.empty? ? nil : child_action.task_queue,
            search_attributes: Temporalio::SearchAttributes._from_proto(proto_sa)
          )
        },
        child_action.awaitable_choice,
        after_started_fn: lambda(&:result),
        after_completed_fn: lambda(&:result)
      )
    when :set_patch_marker
      marker = action.set_patch_marker
      if marker.deprecated
        Temporalio::Workflow.deprecate_patch(marker.patch_id)
        was_patched = true
      else
        was_patched = Temporalio::Workflow.patched(marker.patch_id)
      end
      return handle_action(marker.inner_action) if was_patched
    when :set_workflow_state
      @workflow_state = action.set_workflow_state
    when :await_workflow_state
      target_key = action.await_workflow_state.key
      target_value = action.await_workflow_state.value
      Temporalio::Workflow.wait_condition do
        @workflow_state.kvs.to_h.fetch(target_key, nil) == target_value
      end
    when :upsert_memo
      nil
    when :upsert_search_attributes
      updates = action.upsert_search_attributes.search_attributes.map do |k, v|
        if k.include?('Keyword')
          Temporalio::SearchAttributes::Update.new(
            Temporalio::SearchAttributes::Key.new(k, :keyword),
            v.data[0].to_s
          )
        else
          Temporalio::SearchAttributes::Update.new(
            Temporalio::SearchAttributes::Key.new(k, :int),
            v.data[0].unpack1('q<')
          )
        end
      end
      Temporalio::Workflow.upsert_search_attributes(*updates)
    when :nested_action_set
      return handle_action_set(action.nested_action_set)
    when :nexus_operation
      raise Temporalio::Error::ApplicationError.new(
        'ExecuteNexusOperation is not supported',
        non_retryable: true
      )
    else
      raise Temporalio::Error::ApplicationError, "unrecognized action: #{action}"
    end

    nil
  end

  def launch_activity(exec_activity)
    act_type = 'noop'
    args = []

    case exec_activity.activity_type
    when :delay
      act_type = 'delay'
      args << exec_activity.delay
    when :noop
      act_type = 'noop'
    when :payload
      act_type = 'payload'
      input_data = Array.new(exec_activity.payload.bytes_to_receive) { |i| i % 256 }.pack('C*')
      args << input_data
      args << exec_activity.payload.bytes_to_return
    when :client
      act_type = 'client'
      args << exec_activity.client
    when :retryable_error
      act_type = 'retryable_error'
      args << exec_activity.retryable_error
    when :timeout
      act_type = 'timeout'
      args << exec_activity.timeout
    when :heartbeat
      act_type = 'heartbeat'
      args << exec_activity.heartbeat
    end

    options = {
      schedule_to_close_timeout: duration_or_nil(exec_activity, :schedule_to_close_timeout),
      start_to_close_timeout: duration_or_nil(exec_activity, :start_to_close_timeout),
      schedule_to_start_timeout: duration_or_nil(exec_activity, :schedule_to_start_timeout)
    }.compact

    if exec_activity.retry_policy
      options[:retry_policy] = Temporalio::RetryPolicy.new(
        max_attempts: exec_activity.retry_policy.maximum_attempts,
        initial_interval: exec_activity.retry_policy.initial_interval&.then do |d|
          d.seconds + (d.nanos / 1_000_000_000.0)
        end
      )
    end

    if exec_activity.locality == :is_local
      Temporalio::Workflow.execute_local_activity(act_type, *args, **options)
    else
      if exec_activity.respond_to?(:has_priority?) && exec_activity.has_priority?
        raise NotImplementedError, 'priority is not supported yet'
      end

      ht = duration_or_nil(exec_activity, :heartbeat_timeout)
      options[:heartbeat_timeout] = ht if ht
      options[:task_queue] = exec_activity.task_queue unless exec_activity.task_queue.empty?
      if exec_activity.locality == :remote
        options[:cancellation_type] = convert_cancel_type(exec_activity.remote.cancellation_type)
      end

      Temporalio::Workflow.execute_activity(act_type, *args, **options)
    end
  end

  def handle_awaitable_choice(
    start_fn,
    choice,
    after_started_fn: method(:brief_wait),
    after_completed_fn: method(:wait_task_complete)
  )
    if choice.nil? || choice.condition.nil?
      start_fn.call
      return
    end

    case choice.condition
    when :abandon
      nil
    when :cancel_before_started
      task = Temporalio::Workflow.async { start_fn.call }
      task.cancel
      begin
        task.result
      rescue Temporalio::Error::CancelledError
        nil
      end
    when :cancel_after_started
      task = Temporalio::Workflow.async { start_fn.call }
      after_started_fn.call(task)
      task.cancel
      begin
        task.result
      rescue Temporalio::Error::CancelledError
        nil
      end
    when :cancel_after_completed
      task = Temporalio::Workflow.async { start_fn.call }
      after_completed_fn.call(task)
      task.cancel
    when :wait_finish
      start_fn.call
    end
  end

  def brief_wait(_task)
    Temporalio::Workflow.sleep(0.001)
  end

  def wait_task_complete(task)
    task.result
  end

  def convert_cancel_type(ctype)
    case ctype
    when :TRY_CANCEL then Temporalio::Workflow::ActivityCancellationType::TRY_CANCEL
    when :WAIT_CANCELLATION_COMPLETED then Temporalio::Workflow::ActivityCancellationType::WAIT_CANCELLATION_COMPLETED
    when :ABANDON then Temporalio::Workflow::ActivityCancellationType::ABANDON
    else raise NotImplementedError, "Unknown cancellation type #{ctype}"
    end
  end

  def duration_or_nil(activity_action, field)
    val = activity_action.send(field)
    return nil if val.nil? || (val.seconds.zero? && val.nanos.zero?)

    val.seconds + (val.nanos / 1_000_000_000.0)
  end
end
